### PR TITLE
First wave of audit remediation recommendations

### PR DESCRIPTION
**Pull request by smartcontrart**
Pending integration of approval mechanism

### CODE REVIEW

File: source/swap/contracts/Swap.sol [hunk 1]
```diff
@@ -98,7 +98,7 @@ contract Swap is ISwap, Ownable {
   ) {
     if (_protocolFee >= FEE_DIVISOR) revert InvalidFee();
     if (_protocolFeeWallet == address(0)) revert InvalidFeeWallet();
-    uint256 currentChainId = getChainId();
+    uint256 currentChainId = block.chainid;
     DOMAIN_CHAIN_ID = currentChainId;
```
Comment (dmosites) [inline source/swap/contracts/Swap.sol]: Just combine these two lines and assign `DOMAIN_CHAIN_ID = block.chaind`?

File: source/swap/contracts/Swap.sol [hunk 2]
```diff
@@ -495,8 +496,6 @@ contract Swap is ISwap, Ownable {
     address token,
     bytes4 kind
   ) internal {
-    // Ensure the transfer is not to self.
-    if (from == to) revert SelfTransferInvalid();
```
Comment (dmosites) [inline source/swap/contracts/Swap.sol:485]: Reason to remove this?

File: source/swap/contracts/Swap.sol [hunk 3]
```diff
@@ -522,24 +517,21 @@ contract Swap is ISwap, Ownable {
...
   {
     uint256 groupKey = nonce / 256;
     uint256 indexInGroup = nonce % 256;
     uint256 group = _nonceGroups[signer][groupKey];
 
-    // If it is already used, return false
+    // If it is already used, return mit cancel and revert
     if ((group >> indexInGroup) & 1 == 1) {
-      return false;
+      emit Cancel(nonce, msg.sender);
```
Comment (dmosites) [inline source/swap/contracts/Swap.sol:508]: `Cancel` should only be emitted from the `cancel` function. Would place it after L319.

**Review by dmosites — CHANGES_REQUESTED** [review]

File: source/swap/contracts/Swap.sol (hunk 3 above)
Comment (smartcontrart) [reply source/swap/contracts/Swap.sol:508]: Makes ssenes

File: source/swap/contracts/Swap.sol (hunk 1 above)
Comment (dmosites) [reply source/swap/contracts/Swap.sol]: Sorry, this was commented prior to the EIP712 commit.

File: source/swap/contracts/Swap.sol (hunk 2 above)
Comment (dmosites) [reply source/swap/contracts/Swap.sol:485]: Happy to drop it and save gas.

File: source/swap/contracts/Swap.sol [hunk 4]
```diff
@@ -509,28 +498,22 @@ contract Swap is ISwap, Ownable {
...
-    internal
-    returns (bool)
-  {
+  function _markNonceAsUsed(address signer, uint256 nonce) internal {
     uint256 groupKey = nonce / 256;
     uint256 indexInGroup = nonce % 256;
     uint256 group = _nonceGroups[signer][groupKey];
 
-    // If it is already used, return false
+    // If it is already used, return mit cancel and revert
```
Comment (dmosites) [inline source/swap/contracts/Swap.sol:507]: // If it is already used then revert

File: source/swap/contracts/Swap.sol (hunk 2 above)
Comment (smartcontrart) [reply source/swap/contracts/Swap.sol:485]: Leaving as is then

File: source/swap/contracts/Swap.sol (hunk 4 above)
Comment (smartcontrart) [reply source/swap/contracts/Swap.sol:507]: done

**Review by dmosites — APPROVED** [review]

**Pull request merged**